File: fastify_server/src/addons/lifecycle.addon.ts
import { readdir } from "node:fs/promises";
import { join } from "node:path";
import { pathToFileURL } from "node:url";
import type { Addon } from "../registry/registry.js";
import type { HookFn } from "../contract/index.js";
import { createLoaderReport, reportError, sortByNumericPrefix } from "../contract/index.js";

const LIFECYCLE_SUFFIXES = [".lifecycle.mjs", ".lifecycle.js"];

async function discoverLifecycleFiles(dir: string): Promise<string[]> {
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    const matches: string[] = [];
    for (const entry of entries) {
      if (!entry.isFile()) continue;
      if (LIFECYCLE_SUFFIXES.some((s) => entry.name.endsWith(s))) {
        matches.push(join(dir, entry.name));
      }
    }
    return sortByNumericPrefix(matches);
  } catch (err) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT" || code === "ENOTDIR") return [];
    throw err;
  }
}

interface LifecycleModule {
  default?: unknown;
  onInit?: HookFn;
  onStartup?: HookFn;
  onShutdown?: HookFn;
}

export const lifecycleAddon: Addon = {
  name: "lifecycle",
  priority: 20,
  async run(_server, config, ctx) {
    const report = createLoaderReport("lifecycle");
    let initCount = 0;
    let startCount = 0;
    let stopCount = 0;

    for (const dir of config.paths.lifecycles) {
      let files: string[] = [];
      try {
        files = await discoverLifecycleFiles(dir);
      } catch (err) {
        reportError(report, "discover", err, dir);
        continue;
      }
      report.discovered += files.length;

      for (const file of files) {
        let mod: LifecycleModule;
        try {
          mod = (await import(pathToFileURL(file).href)) as LifecycleModule;
          report.imported += 1;
        } catch (err) {
          reportError(report, "import", err, file);
          continue;
        }

        const source = (mod.default && typeof mod.default === "object")
          ? (mod.default as LifecycleModule)
          : mod;

        let registeredAny = false;
        if (typeof source.onInit === "function") {
          ctx.registerInitHook(source.onInit);
          initCount += 1;
          registeredAny = true;
        }
        if (typeof source.onStartup === "function") {
          ctx.registerStartupHook(source.onStartup);
          startCount += 1;
          registeredAny = true;
        }
        if (typeof source.onShutdown === "function") {
          ctx.registerShutdownHook(source.onShutdown);
          stopCount += 1;
          registeredAny = true;
        }
        if (registeredAny) report.registered += 1;
        else report.skipped += 1;
      }
    }

    report.details = { init_hooks: initCount, startup_hooks: startCount, shutdown_hooks: stopCount };
    return report;
  },
};

export default lifecycleAddon;

File: fastapi_server/fastapi_server/bootstrap/addons/lifecycle_addon.py
from __future__ import annotations

from fastapi_server.bootstrap.contract import create_loader_report, report_error
from fastapi_server.bootstrap.registry.registry import Addon
from fastapi_server.bootstrap.addons._discover import find_matching_files, import_file

_SUFFIXES = (".lifecycle.py",)
_HOOK_KEYS = ("on_init", "on_startup", "on_shutdown")


def _run(_server, config, ctx):
    report = create_loader_report("lifecycle")
    init_count = 0
    start_count = 0
    stop_count = 0

    for d in config.paths.lifecycles:
        try:
            files = find_matching_files(d, _SUFFIXES)
        except Exception as err:  # noqa: BLE001
            report_error(report, "discover", err, d)
            continue
        report.discovered += len(files)

        for file in files:
            try:
                mod = import_file(file, module_name_prefix="polyglot_lifecycle")
                report.imported += 1
            except Exception as err:  # noqa: BLE001
                report_error(report, "import", err, file)
                continue

            registered_any = False
            on_init = getattr(mod, "on_init", None)
            on_startup = getattr(mod, "on_startup", None)
            on_shutdown = getattr(mod, "on_shutdown", None)

            if callable(on_init):
                ctx.register_init_hook(on_init)
                init_count += 1
                registered_any = True
            if callable(on_startup):
                ctx.register_startup_hook(on_startup)
                start_count += 1
                registered_any = True
            if callable(on_shutdown):
                ctx.register_shutdown_hook(on_shutdown)
                stop_count += 1
                registered_any = True

            if registered_any:
                report.registered += 1
            else:
                report.skipped += 1

    report.details = {
        "init_hooks": init_count,
        "startup_hooks": start_count,
        "shutdown_hooks": stop_count,
    }
    return report


lifecycle_addon = Addon(name="lifecycle", priority=20, run=_run)
